docs/chainlink-keepers/flexible-upkeeps.md

```diff
@@ -0,0 +1,157 @@
+
+## Solution: Perform complex computations with no gas fees
+
+In this section, we will slightly modify the contract and move the computation to `checkUpkeep()` function, which <ins>doesn’t consume any gas</ins>. Moreover, we will support multiple upkeeps for the same contract to parallelize the work to be done.
+The main differences with the previous contract are the following (cf. contract code below):
+
+- The `checkUpkeep()` function receives [checkData](/docs/chainlink-keepers/compatible-contracts/#checkdata), which is used to pass arbitrary bytes to the function. In this case, we will pass a `lowerBound` and an `upperBound` to scope the work to a subarray of `balances`. This will allow us to create several upkeeps with different values of `checkData`. The function loops over the subarray and looks for the indexes of the elements which require balancing and calculates the required `increments`. It then returns `upkeepNeeded == true`, and `performData` , which is calculated by encoding `indexes` and `increments`. Note that `checkUpkeep()` is a view function, meaning that any computation does not consume any gas.
+- The `performUpkeep()` function takes [performData](/docs/chainlink-keepers/compatible-contracts/#performdata-1) as a parameter then decodes it to fetch the `indexes` and the `increments`.
+
+> ⚠️ **Note on `performData`**
+>
+> This data should always be validated against the contract’s current state to ensure that `performUpkeep()` is idempotent. It also blocks malicious keepers from sending non-valid data. In this example, we test that the state is correct after rebalancing:
+> `require(_balance == LIMIT, "Provided increment not correct");`
+
+```solidity
+{% include 'samples/Keepers/BalancerOffChain.sol' %}
+```
+
+<div class="remix-callout">
+    <a href="https://remix.ethereum.org/#url=https://docs.chain.link/samples/Keepers/BalancerOffChain.sol" >Open in Remix</a>
+    <a href="/docs/conceptual-overview/#what-is-remix" >What is Remix?</a>
+</div>
+
+Let’s now perform the same test to compare the gas fees:
+
+1. Deploy the contract using Remix on the [supported testnet](../supported-networks) of your choice.
+
+1. Withdraw 100 at 10,100,300,350,500,600,670,700,900. Pass the following to the withdarw function: 100,[10,100,300,350,500,600,670,700,900] (Same test as the [previous section](#problem-on-chain-computation-leads-to-high-gas-fees)).
+
+1. Now, register 3 upkeeps for your contract as explained [here](/docs/chainlink-keepers/register-upkeep/). Because the keepers handle much of the computation off-chain, a gas limit of 200,000 should be sufficient. For each registration, pass the following _CheckData_ values(second column) to specify which balance indexes the registration will monitor. **Note**: Remove any breaking line when copying the values.
```

Comment:
```suggestion
1. Register 3 upkeeps for your contract as explained [here](/docs/chainlink-keepers/register-upkeep/). Because the keepers handle much of the computation off-chain, a gas limit of 200,000 should be sufficient. For each registration, pass the following _CheckData_ values(second column) to specify which balance indexes the registration will monitor. **Note**: Remove any breaking line when copying the values.
```